src/libtest/lib.rs

```diff
@@ -1492,6 +1563,107 @@ fn calc_result(desc: &TestDesc, task_result: Result<(), Box<dyn Any + Send>>) ->
     }
 }
 
+fn run_test_in_process(desc: TestDesc,
+                       nocapture: bool,
+                       testfn: Box<dyn FnOnce() + Send>,
+                       monitor_ch: Sender<MonitorMsg>) {
+    // Buffer for capturing standard I/O
+    let data = Arc::new(Mutex::new(Vec::new()));
+
+    let oldio = if !nocapture {
+        Some((
+            io::set_print(Some(Box::new(Sink(data.clone())))),
+            io::set_panic(Some(Box::new(Sink(data.clone())))),
+        ))
+    } else {
+        None
+    };
+
+    let result = catch_unwind(AssertUnwindSafe(testfn));
+
+    if let Some((printio, panicio)) = oldio {
+        io::set_print(printio);
+        io::set_panic(panicio);
+    }
+
+    let test_result = match result {
+        Ok(()) => calc_result(&desc, Ok(())),
+        Err(e) => calc_result(&desc, Err(e.as_ref())),
+    };
+    let stdout = data.lock().unwrap().to_vec();
+    monitor_ch.send((desc.clone(), test_result, stdout)).unwrap();
+}
+
+fn spawn_test_subprocess(desc: TestDesc, monitor_ch: Sender<MonitorMsg>) {
+    let (result, test_output) = (|| {
+        let args = env::args().collect::<Vec<_>>();
+        let current_exe = &args[0];
+        let output = match Command::new(current_exe)
+            .env(SECONDARY_TEST_INVOKER_VAR, desc.name.as_slice())
+            .output() {
+                Ok(out) => out,
+                Err(e) => {
+                    let err = format!("Failed to spawn {} as child for test: {:?}", args[0], e);
+                    return (TrFailed, err.into_bytes());
+                }
+            };
+
+        let std::process::Output { stdout, stderr, status } = output;
+        let mut test_output = stdout;
+        test_output.extend_from_slice(&stderr);
+
+        let result = match (move || {
+            let exit_code = status.code().ok_or("child process was terminated")?;
+            TestResult::try_from(exit_code).map_err(|_| {
+                format!("child process returned unexpected exit code {}", exit_code)
+            })
+        })() {
+            Ok(r) => r,
+            Err(e) => {
+                write!(&mut test_output, "Unexpected error: {}", e).unwrap();
+                TrFailed
+            }
+        };
+
+        (result, test_output)
+    })();
+
+    monitor_ch.send((desc.clone(), result, test_output)).unwrap();
+}
+
+fn run_test_in_spawned_subprocess(desc: TestDesc, testfn: Box<dyn FnOnce() + Send>) {
```

Comment:
This function could probably `-> !` and avoid the `unreachable!()` further above when it's called

Reply:
heh, I keep forgetting this is stable